consensus/src/quorum_store/batch_reader.rs

```diff
@@ -1,27 +1,485 @@
+        let (tx, rx) = oneshot::channel();
 
-        // TODO: coming soon
+        if let Some(value) = self.db_cache.get(proof.digest()) {
+            if payload_storage_mode(&value) == StorageMode::PersistedOnly {
+                assert!(
+                    value.maybe_payload.is_none(),
+                    "BatchReader payload and storage kind mismatch"
+                );
+                self.batch_store_tx
+                    .send(BatchStoreCommand::BatchRequest(
+                        *proof.digest(),
+                        self.my_peer_id,
+                        Some(tx),
+                    ))
+                    .await
+                    .expect("Failed to send to BatchStore");
+            } else {
+                // Available in memory.
+                if tx
+                    .send(Ok(value
+                        .maybe_payload
+                        .clone()
+                        .expect("BatchReader payload and storage kind mismatch")))
+                    .is_err()
+                {
+                    debug!(
+                        "Receiver of requested batch is not available for digest {}",
+                        proof.digest()
+                    );
+                }
+            }
+        } else {
+            // Quorum store metrics
+            counters::MISSED_BATCHES_COUNT.inc();
 
+            self.self_tx
+                .send(BatchReaderCommand::GetBatchForSelf(proof, tx))
+                .await
+                .expect("Batch Reader Receiver is not available");
+        }
         rx
     }
 
-    pub async fn update_certified_round(&self, _certified_time: LogicalTime) {
-        // TODO: coming soon
+    pub(crate) async fn start<T: QuorumStoreSender + Clone>(
+        &self,
+        mut batch_reader_rx: Receiver<BatchReaderCommand>,
+        network_sender: T,
+        request_num_peers: usize,
+        request_timeout_ms: usize,
+        verifier: ValidatorVerifier,
+    ) {
+        debug!(
+            "[QS worker] BatchReader worker for epoch {} starting",
+            self.epoch()
+        );
+
+        let mut batch_requester = BatchRequester::new(
+            self.epoch(),
+            self.my_peer_id,
+            request_num_peers,
+            request_timeout_ms,
+            network_sender.clone(),
+        );
+
+        let mut interval = time::interval(Duration::from_millis(100));
+
+        loop {
+            tokio::select! {
+                biased;
+
+                _ = interval.tick() => {
+                    batch_requester.handle_timeouts().await;
+                    if self.shutdown_flag.load(Ordering::Relaxed) {
+                        break;
+                    }
+                },
+
+                Some(cmd) = batch_reader_rx.recv() => {
+                    match cmd {
+                        BatchReaderCommand::GetBatchForPeer(digest, peer_id) => {
+                            if let Some(value) = self.db_cache.get(&digest) {
```

Comment:
I think there's an issue here. `db_cache` is entirely in memory, so if all signers restarted at once, and another validator asks for this batch, it'll never get a copy.

To fix this, we'd have to check the DB even when there is no entry in `db_cache`. It will allow a malicious validator to keep requesting batches and having to hit DB though.

WDYT @sasha8 @gelash ?

Reply:
I think we actually thought about it. When a validator restarts it rebuilds the hashmap from storage. We read from storage in BatchStore::new and then pass the db_conetnt to Batch_reader to reconstruct the in-memory db_cache.